blurring_as_a_service/cleaning_pipeline/components/smart_sampling.py

```diff
@@ -55,37 +76,337 @@ def smart_sampling(
+
+    Returns
+    -------
+    Tuple[Dict[str, List[str]], List[str]]
+        A tuple containing two elements:
+        - A dictionary where keys are bin labels and values are lists of image identifiers in that bin.
+        - A list of bin labels.
+
+    If no detections are found, an empty dictionary and list are returned.
+    """
+    
+    if not image_counts:
+        logger.info("No detections found for the given criteria.")
+        return {}, []
+
+    # Calculate min and max counts
+    counts = image_counts.values()
+    min_count, max_count = min(counts), max(counts)
+    logger.info(f"Minimum number of detections for an image: {min_count}")
+    logger.info(f"Maximum number of detections for an image: {max_count}")
+
+    # Determine the range and define bin size strategy
+    detection_range = max_count - min_count
+    bin_size = determine_bin_size(detection_range)
+
+    # Calculate the bin edges
+    bins = np.linspace(min_count, max_count, bin_size + 1)
+
+    # Initialize a dictionary to hold bin counts
+    bin_counts, bin_labels = initialize_bin_counts(bins)
+
+    # Categorize images into bins
+    categorize_into_bins(image_counts, bins, bin_labels, bin_counts)
+
+    return bin_counts, bin_labels
+
+def determine_bin_size(detection_range: int) -> int:
+    """
+    Determines the bin size for categorization based on the detection range.
+
+    Parameters
+    ----------
+    detection_range : int
+        The range of detection counts across all images.
+
+    Returns
+    -------
+    int
+        The number of bins to be used for categorization.
+    """
+    
+    if detection_range <= 10:
+        return 3
+    elif 10 < detection_range <= 50:
+        return 5
+    else:
+        return 10
+
+def initialize_bin_counts(bins: np.ndarray) -> Tuple[Dict[str, List], List[str]]:
+    """
+    Initializes the counts and labels for bins.
+
+    Parameters
+    ----------
+    bins : np.ndarray
+        The array of bin edges.
+
+    Returns
+    -------
+    Tuple[Dict[str, List], List[str]]
+        A tuple containing two elements:
+        - A dictionary where keys are bin labels and values are empty lists for each bin.
+        - A list of bin labels.
+    """
+    
+    bin_counts = {}
+    bin_labels = [f"{int(bins[i])}-{int(bins[i + 1]) - 1}" for i in range(len(bins) - 1)]
+    for label in bin_labels:
+        bin_counts[label] = []
+    return bin_counts, bin_labels
+
+def categorize_into_bins(image_counts: Dict[str, int], bins: np.ndarray, bin_labels: List[str], bin_counts: Dict[str, List[str]]) -> None:
+    """
+    Categorizes each image into a bin based on its detection count.
+
+    Parameters
+    ----------
+    image_counts : Dict[str, int]
+        A dictionary mapping image identifiers to their detection counts.
+    bins : np.ndarray
+        The array of bin edges.
+    bin_labels : List[str]
+        A list of bin labels.
```

Comment:
What are the bin labels? Can you make me an example?

Reply:
Each bin label is a `string` descriptor of the range of detection counts that belongs to each respective bin. When images are processed and their detection counts are categorized, they are placed into these bins based on where their count falls in these ranges. 

For example, if we have a range of detection counts from 0 to 100, we decide to create bins like [0-9], [10-19], ...., [90-99], [100]. The text representing each range (like "0-9" or "10-19") is a bin label. In this example, an image with 15 detections would fall into the bin labeled "10-19".